src/python/pants/base/specs.py

```diff
@@ -201,14 +201,14 @@ def make_glob_patterns(self, address_mapper):
 }
 
 
-def more_specific(spec1: Spec, spec2: Spec) -> Spec:
+def more_specific(spec1: Optional[Spec], spec2: Spec) -> Spec:
   """Returns which of the two specs is more specific.
 
   This is useful when a target matches multiple specs, and we want to associate it with
   the "most specific" one, which will make the most intuitive sense to the user.
   """
   # Note that if either of spec1 or spec2 is None, the other will be returned.
```

Comment:
This function looks broken in the small: What if both are `None`? What if spec2 is `None`? I realize you're just trying to get typechecking to pass, but this probably deserves a separate PR to fix first.

Reply:
See https://github.com/pantsbuild/pants/pull/8639#discussion_r347729508. I agree the function should fail if both are `None`. Also, `spec_test.py` (which isn't yet type checked) treats both args as `Optional`.

Reply:
As I mentioned in the comment Eric links, this was an attempt to use type-checking to enforce the types that were already used in non-test code, in an attempt to avoid having to code more defensively here. It's easy to make the change to explicitly check for both being None.